04-onion-routing.md

```diff
@@ -248,18 +248,111 @@ This is a more flexible format, which avoids the redundant `short_channel_id` fi
+    - MUST use the legacy payload format instead.
+  - For every node:
+    - MUST include `amt_to_forward` and `outgoing_cltv_value`.
+  - For every non-final node:
+    - MUST include `short_channel_id`
+	- MUST NOT include `payment_data`
+  - For the final node:
+    - MUST NOT include `short_channel_id`
+    - if the recipient provided `payment_secret`:
+      - MUST include `payment_data`
+	  - MUST set `payment_secret` to the one provided
+	  - MUST set `total_msat` to the total amount it will send
 
 The reader:
   - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
+  - if it is the final node:
+    - MUST treat `total_msat` as if it were equal to `amt_to_forward` if it
+	  is not present.
+
+The requirements for the contents of these fields are specified [above](#legacy-hop_data-payload-format)
+and [below](#basic-multi-part-payments).
+
+### Basic Multi-Part Payments
+
+An HTLC may be part of a larger "multi-part" payment: such
+"base" atomic multipath payments will use the same `payment_hash` for
+all paths.
+
+Note that `amt_to_forward` is the amount for this HTLC only: a
+`total_msat` field containing a greater value is a promise by the
+ultimate sender that the rest of the payment will follow in succeeding
+HTLCs; we call these outstanding HTLCs which have the same preimage,
+an "HTLC set".
+
+#### Requirements
 
-The requirements for the contents of these fields are specified [above](#legacy-hop_data-payload-format).
+The writer:
+  - if the invoice offers the `basic_mpp` feature:
+    - MAY send more than one HTLC to pay the invoice.
+    - MUST use the same `payment_hash` on all HTLCs in the set.
+    - SHOULD send all payments at approximately the same time.
+    - SHOULD try to use diverse paths to the recipient for each HTLC.
+    - SHOULD retry and/or re-divide HTLCs which fail.
+    - if the invoice specifies an `amount`:
+       - MUST set `total_msat` to at least that `amount`, and less
+         than or equal to twice `amount`.
+    - otherwise:
+      - MUST set `total_msat` to the amount it wishes to pay.
+    - MUST ensure that the total `amount_msat` of the HTLC set which arrives at the payee
+      is equal to `total_msat`.
+    - MUST NOT send another HTLC if the total `amount_msat` of the HTLC set is already greater or equal to `total_msat`.
+  - otherwise:
+    - MUST set `total_msat` equal to `amt_to_forward`.
+
+The final node:
+  - MUST fail the HTLC if dictated by Requirements under [Failure Messages](#failure-messages)
+    - Note: "amount paid" specified there is the `total_msat` field.
+  - if it does not support `basic_mpp`:
+    - MUST fail the HTLC if `total_msat` is not exactly equal to `amt_to_forward`.
+  - otherwise, if it supports `basic_mpp`:
+    - MUST add it to the HTLC set corresponding to that `payment_hash`.
+    - SHOULD fail the entire HTLC set if `total_msat` is not the same for
```

Comment:
another thing that came up for us when implementing the receiver logic: let's say one mpp HTLC has arrived, but is lower than `total_msat`. then a legacy payment arrives settling the full amount, should the receiver settle the legacy payment and cancel the mpp HTLC? or keep waiting for mpp?

it could be that the payment_secret is optional and the sender doesn't understand it, but rejecting the legacy means it's actually compulsory. tbh this situation only seems totally resolvable once it is fully required, but i wonder, is one of the approaches better than the other?

Reply:
Honestly this looks like something that will never happen: IIUC you're assuming that two different people are trying to pay the same invoice, right?
Because if Alice sent the first mpp HTLC, there's no reason for her to send a legacy payment next. So it has to come from someone else?

If that's the scenario you're exploring, I believe the legacy payment "wins". It's the first payment that pays the full amount, so it should receive the preimage.

I think this raises an interesting point about proof-of-payment. If multiple people are collaborating to pay an invoice with MPP, they currently all receive the preimage, but they didn't pay the whole thing. The proof-of-payment is now weaker than before: it only says that you paid part of that invoice. Is that a problem? Not sure.

Reply:
> Because if Alice sent the first mpp HTLC, there's no reason for her to send a legacy payment next. So it has to come from someone else?

An intermediary can see that the first MPP HTLC is stuck being held, then start probing with regular payments. In the case of a zero value invoice, they could settle the invoice for a value much lower than `total_msat`, then settle back and take the difference.

Until we have mandatory payment secrets, I think it's actually safer to reject legacy payments if you have already started constructing an MPP HTLC set, that's how we have it implemented in our draft. Allowing legacy to override would reintroduce the probing vector afaict.

Reply:
> I think this raises an interesting point about proof-of-payment. If multiple people are collaborating to pay an invoice with MPP, they currently all receive the preimage, but they didn't pay the whole thing. The proof-of-payment is now weaker than before: it only says that you paid part of that invoice. Is that a problem? Not sure.

More evidence that we should stop calling preimages proofs of payment, because they aren't :)

Reply:
> An intermediary can see that the first MPP HTLC is stuck being held, then start probing with regular payments.

Got it. However, I think there's another way to resolve that. In my opinion if the client is doing MPP (sending multiple shards), then it must include the payment secret. This prevents the probing you're mentioning.
And in that case the receiver simply needs to enforce that other HTLCs for that payment_hash include the payment_secret.

It's really hard to express correctly the intent with feature bits though. In my mind, when the invoice contains `basic_mpp_optional || payment_secret_optional`, this means:

- I allow legacy payments (single HTLC without `payment_secret`)
- Or MPP payments but in that case the `payment_secret` is required

WDYT?

Reply:
This is fixed by requiring the payment secret (see #712 which is a separate PR since this was already accepted at the meeting).

In this case, if the legacy payment arrives first, it's accepted (assuming that you haven't made payment_secret compulsory yet!).  If mpp arrives first, the secret requirement is now locked in, which is a security win.

> More evidence that we should stop calling preimages proofs of payment, because they aren't :)

They absolutely *are*.  They are proof that payment was made.  They *don't* prove who paid it, which indeed sucketh!

Reply:
> This is fixed by requiring the payment secret (see #712 which is a separate PR since this was already accepted at the meeting).

Chatted on IRC w/ @rustyrussell, indeed #712 prevents this along with the receiver rejecting legacy payments if they advertised required payment secrets (which is already in this pr).